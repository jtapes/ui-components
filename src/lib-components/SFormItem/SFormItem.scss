@import '../../theme/styles/variables';
@import '../../theme/styles/mixins';
@import '../../theme/styles/reset';
@import '../../theme/styles/fonts';
@import '../../theme/styles/global';

.s-form-item {
  &__error {
    @include text-sm;

    margin-top: $size--8;
    color: $color--danger;
  }

  &--required {
    .s-input {
      &__placeholder {
        &::after {
          content: '*';
          display: inline-block;
          color: $color--danger;
        }
      }
    }
  }
}
