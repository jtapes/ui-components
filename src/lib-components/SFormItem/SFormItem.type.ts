export type SlotType = {
  name: string
  content?: string
  icon?: {
    name: string
    class: string
  }
  callback?: Function
}

export type FormItemRulesType = {
  [key: string]: undefined | number | string | number[] | string[]
}
