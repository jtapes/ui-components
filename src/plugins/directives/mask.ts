import Vue from 'vue'
import { DirectiveBinding } from 'vue/types/options'

export type TokenType = {
  [key: string]: {
    pattern?: RegExp
    repeat?: boolean
  }
}
export type MaskType = string | { mask: string; tokens: TokenType }

interface MaskedElement extends HTMLInputElement {
  mask?: MaskType
}

const TOKENS = {
  '#': { pattern: /[0-9]/ },
  x: { pattern: /[0-9a-zA-Z]/ },
  X: { pattern: /[0-9а-яА-Яa-zA-Z]/ },
  S: { pattern: /[a-zA-Z]/ },
  E: { pattern: /[0-9a-zA-Z-_@.]/ },
  '*': { repeat: true },
}

function createEvent(name: string) {
  const event = document.createEvent('Event')
  event.initEvent(name, true, true)

  return event
}

function getInputElement(el: HTMLElement) {
  return el instanceof HTMLInputElement ? el : el.querySelector('input')
}

function masking(value: string, mask: string, tokens: TokenType): string {
  let iValue = 0
  let iMask = 0
  let output = ''

  while (iValue < value.length && iMask < mask.length) {
    const token = mask[iMask]
    const maskRegex = tokens[token]
    const currentValue = value[iValue]

    if (maskRegex && maskRegex.pattern) {
      if (maskRegex.pattern.test(currentValue)) {
        output += currentValue
        iMask++
      }

      iValue++
    } else if (maskRegex && maskRegex.repeat) {
      const prevToken = mask[iMask - 1]
      const prevMaskRegex = tokens[prevToken]

      if (prevMaskRegex && prevMaskRegex.pattern) {
        if (!prevMaskRegex.pattern.test(currentValue)) {
          iMask++
        } else {
          iMask--
        }
      }
    } else {
      output += token
      if (currentValue === token) iValue++

      iMask++
    }
  }

  return output
}

function handleValueChange(e: Event) {
  const target = e.target as MaskedElement
  const { mask } = target
  const oldValue = target.value
  if (mask) {
    if (typeof mask === 'string') {
      target.value = masking(oldValue, mask, TOKENS)
    } else {
      target.value = masking(oldValue, mask.mask, mask.tokens)
    }

    if (target.value !== oldValue) {
      target.dispatchEvent(createEvent('input'))
    }
  }
}

function toggleMask(el: HTMLElement, binding: DirectiveBinding) {
  const input = getInputElement(el) as MaskedElement

  if (input && binding.value) {
    input.mask = binding.value
    input.addEventListener('input', handleValueChange)

    if (input.value) {
      input.dispatchEvent(createEvent('input'))
    }
  }
}

export default () => {
  Vue.directive('mask', {
    update(el, binding) {
      if (binding.value && binding.value !== binding.oldValue) {
        Vue.nextTick(() => {
          toggleMask(el, binding)
        })
      }
    },

    bind(el, binding) {
      if (binding.value) {
        toggleMask(el, binding)
      }
    },

    unbind(el, binding) {
      if (binding.value) {
        const input = getInputElement(el)
        if (input) {
          input.removeEventListener('input', handleValueChange)
        }
      }
    },
  })
}
