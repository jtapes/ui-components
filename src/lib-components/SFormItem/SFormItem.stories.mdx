import { config, withDesign } from 'storybook-addon-designs'
import { withTests } from '@storybook/addon-jest'
import { styled } from '@storybook/theming'
import {
  Meta,
  Story,
  Description,
  Source,
  Preview,
} from '@storybook/addon-docs'
import SFormItem, {
  COMPONENT_TYPES,
} from '@/lib-components/SFormItem/SFormItem.vue'
import results from '/.jest-test-results.json'
import mask from '@/plugins/directives/mask'
import Vue from 'vue'

<Meta
  title="DESIGN SYSTEM/Atom/FormItem"
  component={SFormItem}
  decorators={[withDesign, withTests({ results })]}
  argTypes={{
    component: {
      type: { name: 'string', required: true },
      defaultValue: 's-input',
      description: 'Тип компонента',
      control: {
        type: 'radio',
        options: COMPONENT_TYPES,
      },
    },
      isDirty: {
          control: 'boolean',
          defaultValue: true,
      },
    isValid: {
      type: { name: 'boolean', required: false },
      description: 'Для получения статуса валидации',
      control: 'boolean',
      defaultValue: true,
    },
    rules: {
      description: 'Схема для валидации',
      control: 'array',
      defaultValue: [{ name: 'required '}]
    },
    value: { control: 'string' },
  }}
  parameters={{
    design: config({
      type: 'figma',
      url:
        'https://www.figma.com/file/42Xh2qI3qvq2CRkc2HJlwA/MVP-3-%2F-Market-%2F-Catalog-%E2%80%A2-Products-Page?node-id=132%3A0',
    }),
    docs: {
      source: {
        type: 'code',
      },
    },
  }}
/>

# Поле формы

<Description of={SFormItem} />

export const Template = (args, { argTypes }) => {
  Vue.use(mask)
  return {
    props: Object.keys(argTypes),
    components: { SFormItem },
    template: `
      <SFormItem
        v-model="model"
        component="s-input"
        :rules="rules"
        v-bind="$props"
      />
    `,
    data() {
      return {
        model: '',
      }
    },
  }
}

<Preview withToolbar withSource="none">
  <Story
    name="SFormItem"
    args={{
      component: 's-input',
      value: '',
      rules: [{ name: 'required' }],
    }}
  >
    {Template.bind({})}
  </Story>
</Preview>

## Пример

<Source
  language="html"
  code={`
<s-form-item
  v-model="model"
  component="s-input"
  :is-valid.sync="isValid"
  :rules="[{ name: 'required' }]"
  :clearable="true"
  placeholder="Введите информацию"
/>
  `}
/>

<ArgsTable of={SFormItem} />

export const StyledTd = styled.td`
  border: 1px solid rgba(0, 0, 0, 0.1);
  color: #333333;
  margin: 0;
  padding: 6px 13px;
`

export const StyledTh = styled.th`
  font-weight: bold;
  color: #333333;
  border: 1px solid rgba(0, 0, 0, 0.1);
  margin: 0;
  padding: 6px 13px;
`

export const StyledTr = styled.tr`
  border-top: 1px solid rgba(0, 0, 0, 0.1);
  margin: 0;
  padding: 0;
`

export const StyledTable = styled.table`
  margin: 0;
  -webkit-font-smoothing: antialiased;
  -moz-osx-font-smoothing: grayscale;
  -webkit-tap-highlight-color: rgba(0, 0, 0, 0);
  -webkit-overflow-scrolling: touch;
  margin: 16px 0;
  font-size: 14px;
  line-height: 24px;
  padding: 0;
  border-collapse: collapse;
`

## Валидация

<Canvas withSource="none">
  <div>
    Пример:
    <Source
      language="html"
      code={`
<s-form-item
  v-model="form.email"
  class="mt-16"
  placeholder="Email *"
  clearable
  :rules="[
    {
      name: 'regularkaName',
      rule: /^\\d{7}$/g,
    },
    { name: 'minLength', params: [4] },
    {
      name: 'myCustomRule',
      rule: (val) => val === 'test',
    },
    {
      name: 'unionRuleName',
      modifier: 'or',
      rules: [{ rule: /^\\d{7}$/g }, { name: 'minLength', params: [4] }],
    },
  ]"
>
  <template #error-regularkaName="{ error }">
    <div>{{ error.isError ? 'Регулярное выражение!' : '' }}</div>
  </template>
</s-form-item>
      `}
    />
    Правила:
    <StyledTable>
      <tbody>
        {Object.keys(rulesList).map((el, i) => (
          <StyledTr key={i}>
            <StyledTd>{el}</StyledTd>
          </StyledTr>
        ))}
      </tbody>
    </StyledTable>
    Также при необходимости можно использовать один из 3 логических
    модификаторов - not\or\and:
    <br />
    <br />
    NOT:
    <Source
      language="js"
      code={`
{
  name: 'notMobile',
  modifier: 'not',
  rule: /^+7\\d{10}$/,
},
      `}
    />
    OR:
    <Source
      language="js"
      code={`
{
  name: 'unionRuleName',
  modifier: 'or',
  rules: [{ rule: /^\\d{7}$/g }, { name: 'minLength', params: [4] }],
},
      `}
    />
    AND:
    <Source
      language="js"
      code={`
{
  name: 'unionRuleName',
  modifier: 'and',
  rules: [{ rule: /^\\d{7}$/g }, { name: 'minLength', params: [4] }],
},
      `}
    />
    <br />
    Для работы с ошибками мы используем слоты #errors или #error-[имяПравила].
    Первый дает объект со всеми ошибками, в котором можно обратиться по
    названию, второй даёт доступ к конкретной ошибке. При этом если мы
    используем #errors, то он затирает все #error-[имяПравила] слоты. Пример
    объекта с ошибками:
    <Source
      language="js"
      code={`
{
  required: {
    isError: true, // Текущее состояние проверки поля по этому правилу
    msg: 'Поле необходимо для заполнения', // Сообщение при ошибке
    params: {}, // Дополнительные параметры
  },
  minLength: {
    isError: false,
    msg: 'Поле должно иметь минимум 4 символа',
    params: {
      min: 4,
    },
  },
}
      `}
    />
    Пример использования этого объекта:
    <Source
      language="html"
      code={`
<template #error-minLength="{ error }">
  {{ error.isError ? \`Сообщение должно быть не короче \$\{error.params.min\}\` символов! : '' }}
</template>
      `}
    />
  </div>
</Canvas>
