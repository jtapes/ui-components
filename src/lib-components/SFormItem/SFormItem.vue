<template>
  <validation-provider
    v-slot="{ validated, valid, dirty, errors }"
    :name="name"
    :mode="validationMode"
    :immediate="immediate"
    tag="div"
    class="s-form-item"
    :rules="rules"
    :class="{ 's-form-item--required' : isObject(rules) && rules.hasOwnProperty('required') }"
  >
    <component
      :is="component"
      v-mask="mask"
      :value="value"
      v-bind="$attrs"
      :state="(validated || dirty) && !valid ? 'error' : $attrs.state"
      @change="handleValueChange"
    >
      <template v-for="(slot, index) in slots" #[slot.name]>
        <div
          :key="index"
          @click="slot.callback ? slot.callback : false"
          v-html="slot.content"
        />
        <svg-icon
          v-if="slot.icon"
          :key="`svg-${index}`"
          :name="slot.icon.name"
          :class="slot.icon.class"
          @click="slot.callback ? slot.callback : false"
        />
      </template>
    </component>
    <div
      v-if="errors.length"
      class="s-form-item__error"
    >
      <!-- @slot Слот для ошибок -->
      <slot name="errors" :errors="errors">
        <div v-for="(error, errorIndex) in errors" :key="`error-${errorIndex}`">
          <!-- @slot Слот для конкретной ошибки -->
          <slot v-if="error" :name="`error-${errorIndex}`" :error="error">
            {{ error }}
          </slot>
        </div>
      </slot>
    </div>
  </validation-provider>
</template>

<script lang="ts">
import {
  Component,
  Prop,
  Model,
  Vue,
} from 'vue-property-decorator'

import SAutocomplete from '@/lib-components/SAutocomplete/SAutocomplete.vue'
import SInput from '@/lib-components/SInput/SInput.vue'
import SCheckbox from '@/lib-components/SCheckbox/SCheckbox.vue'
import SCheckboxGroup from '@/lib-components/SCheckboxGroup/SCheckboxGroup.vue'
import SRadioGroup from '@/lib-components/SRadioGroup/SRadioGroup.vue'
import SSelect from '@/lib-components/SSelect/SSelect.vue'
import SRadio from '@/lib-components/SRadio/SRadio.vue'
import SDatePicker from '@/lib-components/SDatePicker/SDatePicker.vue'
import SPriceRange from '@/lib-components/SPriceRange/SPriceRange.vue'
import SvgIcon from '@/lib-components/SvgIcon/SvgIcon.vue'
import { MaskType } from '@/plugins/directives/mask'
import { isObject } from '@/utils/utils'

import { extend } from 'vee-validate'
import locale from './validation/locales/ru.json'
import { ValidationProvider } from 'vee-validate'
import * as validationRules from 'vee-validate/dist/rules'

for (const [rule, validation] of Object.entries(validationRules)) {
  extend(rule, {
    ...validation,
    message: (locale.messages as { [key: string]: any })[rule],
  })
}

import type { FormItemRulesType, SlotType } from '@/lib-components/SFormItem/SFormItem.type'

export const COMPONENT_TYPES = [
  's-input',
  's-checkbox',
  's-checkbox-group',
  's-select',
  's-radio',
  's-radio-group',
  's-price-range',
  's-date-picker',
  's-autocomplete'
]

@Component({
  components: {
    SAutocomplete,
    SInput,
    SCheckbox,
    SCheckboxGroup,
    SSelect,
    SRadio,
    SPriceRange,
    SDatePicker,
    SRadioGroup,
    SvgIcon,
    ValidationProvider
  }
})
export default class SFormItem extends Vue {
  @Prop({
    type: String,
    validator(value: any): boolean {
      return COMPONENT_TYPES.includes(value)
    },
  })
  component!: string

  @Model('change')
  value!: any

  @Prop({ type: String, default: ' ' })
  name!: string

  @Prop({ type: String, default: 'eager' })
  validationMode!: string

  @Prop({ type: Boolean, default: false })
  immediate!: boolean

  @Prop(Object)
  rules?: FormItemRulesType

  @Prop()
  mask?: MaskType

  @Prop({ type: Array, default: () => [] })
  slots!: SlotType[]

  handleValueChange(value: any): void {
    this.$emit('change', value)
  }

  isObject = isObject
}
</script>
