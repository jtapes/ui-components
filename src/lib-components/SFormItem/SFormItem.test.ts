import { mount } from '@vue/test-utils'
import Vue from 'vue'
import SFormItem from '@/lib-components/SFormItem/SFormItem.vue'
import mask from '@/plugins/directives/mask'

Vue.use(mask)

describe('SFormItem', () => {
  test('Is Vue snapshot', () => {
    const wrapper = mount(SFormItem, {
      propsData: {
        component: 's-input',
      },
    })

    expect(wrapper.element).toMatchSnapshot()
  })
})
