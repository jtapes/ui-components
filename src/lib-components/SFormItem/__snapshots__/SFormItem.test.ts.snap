// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`SFormItem Is Vue snapshot 1`] = `
<div
  class="s-form-item"
>
  <div
    class="s-input s-input--text s-input--rounded"
  >
    <!---->
     
    <input
      class="s-input__inner"
      type="text"
    />
     
    <!---->
     
    <!---->
     
    <!---->
  </div>
   
  <!---->
</div>
`;
